### PR TITLE
Add `lint-go-gopls`

### PR DESCRIPTION
**Pull request by silverwind**
Uses `gopls check <files>` as a linter. Tested locally and brings up 149 errors currently for me. I don't think I want to fix them in this PR, but I would like at least to get this analysis running on CI.

List of errors:
```
modules/indexer/code/indexer.go:181:11: impossible condition: nil != nil
routers/private/hook_post_receive.go:120:15: tautological condition: nil == nil
services/auth/source/oauth2/providers.go:185:9: tautological condition: nil == nil
services/convert/issue.go:216:11: tautological condition: non-nil != nil
tests/integration/git_test.go:332:9: impossible condition: nil != nil
services/migrations/migrate.go:179:24-43: unused parameter: ctx
services/repository/transfer.go:288:48-69: unused parameter: doer
tests/integration/api_repo_tags_test.go:75:41-61: unused parameter: session
tests/integration/git_test.go:696:64-74: unused parameter: baseBranch
tests/integration/gpg_git_test.go:265:27-39: unused parameter: t
tests/integration/gpg_git_test.go:284:23-29: unused parameter: tmpDir
tests/integration/gpg_git_test.go:284:31-35: unused parameter: name
tests/integration/gpg_git_test.go:284:37-42: unused parameter: email
```

### CODE REVIEW

**Comment by silverwind** [comment]
The command currently [does not set a proper exit code ](https://github.com/golang/go/issues/67078)when there are errors, so we would likely have to analyze the output.

**Comment by silverwind** [comment]
Lint script added which performs error detection and error counting. It should now fail the CI.

File: tools/lint-go-gopls.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+set -euo pipefail
+
+# lint all go files with 'gopls check' and look for lines starting with the
+# current absolute path, indicating a error was found. This is neccessary
+# because the tool does not set non-zero exit code when errors are found.
+# ref: https://github.com/golang/go/issues/67078
+ERROR_LINES=$("$GO" run "$GOPLS_PACKAGE" check $@ | grep -P "^$PWD");
```
Comment (silverwind) [inline tools/lint-go-gopls.sh]:
```suggestion
ERROR_LINES=$("$GO" run "$GOPLS_PACKAGE" check $@ | grep -E "^$PWD");
```

**Comment by silverwind** [comment]
Failure detection looks good, I will now add `|| true` so we can merge it without fixing the errors.

**Review by delvh — APPROVED** [review]

**Comment by silverwind** [comment]
This causes a number of annotations to show up on PRs that we should fix first before merging, otherwise it would be annoying having those show up on every PR.

**Comment by silverwind** [comment]
49 errors left, https://github.com/go-gitea/gitea/pull/30735 will fix a few.

**Comment by silverwind** [comment]
It seems gopls has [no support](https://github.com/golang/go/issues/50764) for `//nolint` currently, so we need to get more creative for technical debt like this one:

https://github.com/go-gitea/gitea/blob/a21ca9b5a5e85d2734bd4cd9e308800eda41a524/modules/graceful/manager_windows.go#L44-L45

**Comment by silverwind** [comment]
It seems all deprecations that are warned here already have `//nolint` comments, so I'll just ignore any error with "is deprected". Some of these should defintely be looked at though.

**Comment by silverwind** [comment]
This is ready now. For reference, here are the remaining `deprecated` errors that are ignored:

```
routers/api/v1/repo/branch.go:237:16-33: opt.OldBranchName is deprecated: true Name of the old branch to create from
routers/api/v1/repo/branch.go:238:37-54: opt.OldBranchName is deprecated: true Name of the old branch to create from
routers/api/v1/repo/branch.go:239:54-71: opt.OldBranchName is deprecated: true Name of the old branch to create from
routers/api/v1/repo/branch.go:520:14-29: form.BranchName is deprecated: true
modules/markup/common/footnote.go:200:13-29: util.FindClosure is deprecated: This function can not handle newlines. Many elements can be existed over multiple lines(e.g. link labels). Use text.Reader.FindClosure.
modules/markup/common/footnote.go:290:13-29: util.FindClosure is deprecated: This function can not handle newlines. Many elements can be existed over multiple lines(e.g. link labels). Use text.Reader.FindClosure.
models/db/sql_postgres_with_schema.go:42:25-38: driver.Execer is deprecated: Drivers should implement [ExecerContext] instead.
models/db/sql_postgres_with_schema.go:67:11-20: stmt.Exec is deprecated: Drivers should implement StmtExecContext instead (or additionally).
```

**Review by lunny — APPROVED** [review]

**Pull request merged**